lib/plugins/aws/provider.js

```diff
@@ -1698,6 +1744,170 @@ class AwsProvider {
       return this.getStackResources(res.NextToken, allResources);
     });
   }
+
+  getEcrRepositoryName() {
+    const serviceName = this.serverless.service.getServiceName();
+    const stage = this.getStage();
+    return `serverless-${serviceName}-${stage}`;
+  }
 }
 
+Object.defineProperties(
+  AwsProvider.prototype,
+  memoizeeMethods({
+    dockerLoginToEcr: d(
+      // TODO: REFACTORING
+      async function () {
+        const registryId = await this.getAccountId();
+
+        // TODO: Error handling?
+        const result = await this.request('ECR', 'getAuthorizationToken', {
+          registryIds: [registryId],
+        });
+        const { authorizationToken, proxyEndpoint } = result.authorizationData[0];
+        const decodedAuthToken = Buffer.from(authorizationToken, 'base64').toString().split(':')[1];
+        const dockerArgs = ['login', '--username', 'AWS', '--password-stdin', proxyEndpoint];
+        const dockerLoginPromise = spawnExt('docker', dockerArgs, {
+          stdio: ['pipe', 'inherit', 'inherit'],
+        });
+        dockerLoginPromise.child.stdin.write(decodedAuthToken);
+        dockerLoginPromise.child.stdin.end();
+        await dockerLoginPromise;
+        // TODO: Better information about how credentials are stored?
+      },
+      { promise: true }
+    ),
+    getOrCreateEcrRepository: d(
+      async function () {
+        const repositoryName = this.getEcrRepositoryName();
+        let repositoryUri;
+        try {
+          const result = await this.request('ECR', 'describeRepositories', {
+            repositoryNames: [repositoryName],
+          });
+          repositoryUri = result.repositories[0].repositoryUri;
+        } catch (err) {
+          if (err.providerError && err.providerError.code === 'RepositoryNotFoundException') {
+            // TODO: CONSIDER MORE PARAMS FOR REPO CREATION?
+            const result = await this.request('ECR', 'createRepository', {
+              repositoryName,
+            });
+            repositoryUri = result.repository.repositoryUri;
+          } else {
+            throw err;
```

Comment:
Style hint.

It usually feels more concise to construct flow as:

```javascript
if (!x) throw error;
... process ...
```

Instead of
```javascript
if (x) {
   ... process ...
} else {
  throw error;
}

Reply:
Thanks @medikoo for all great points raised

> > Should we try to remove ECR repository only if there are provider.images defined?
> 
> Great question. I would attempt to remove repository unconditionally as a whole only when we're removing the service via `sls remove`, still I would attempt it in a way in which it doesn't prolong time of `sls remove` when repository doesn't exist (I proposed how we can do it in one of the comments

I really like the proposed idea :+1: 

> > Do you think it's still worth attempting to re-tag the image in case of failure?
> 
> I think we may either:
> 
> * Skip it, and do not attempt to search and remove unused images
> * Always be clean, and always attempt to leave clean slate (attempt to remove any not used images)
>
> Second feels a bit aggressive, I think we may skip it for now, and eventually revisit if concern would be raised. Working with specific use case may give us better idea on what approach would be most optimal, so we may wait for that.

The only thing I'm worried about here is the potential leftover images generating extra costs for users. Storage in ECR is $0.1/GB/month, which can quickly add up to a significant sum after some time. But I like the idea of potentially revisiting the approach after we get feedback from the community. With that in mind, do you suggest to remove the logic that cleans up unused images after each deploy for now?

> > I don't think we can "fix" it other than enforcing serial build process. Do you think it's fine to keep the outputs like that?
> 
> I'd rather fix it. Outputs like that are very confusing and may lead to wrong assumptions. I'd fix it by flushing output of processes after they finalize and not on the go

Yeah, that's the only way to ensure we don't mix up outputs. The only downside is that users won't see progress of build/push in real-time, but I guess we have to make a tradeoff here. 

> > I was thinking that maybe, we should create an "umbrella" property on provider like `ecr` or `docker`
> 
> Very good call. I propose to use `ecr` (AWS pointed to us once that ECR is not only for `docker`, in light of that `docker` doesn't seem as valid choice)

:+1: 

> > What is the "recommended" way of handling errors from spawnExt? Should we try to catch them and provide a custom error message for user or just fail whenever process fails?
> 
> I think we should catch, and crash with meaningful `ServerlessError`. If we pass through original error, it'll be assumed as _programmer_ error (so like `500`), and will be presented to user with full stack trace (and such errors unconditionally are treated as bugs in our logic).

Makes perfect sense, I'll adjust it

> > Right now it will fail on build attepmt, but I was thinking that maybe we should fail faster in such case.
> 
> If from build fail we can easily diagnose error signals that `docker` is not installed. I would keep that, and just throw meaningful `ServerlessError` in response.

Will try to do that, thanks for a suggestion :+1: 
 
> It'll be great to also test this on Windows. Do you have some access to Windows machine?

Yes, I've planned to try to test it out on Windows machine after wrapping up the development. 

Thanks once again for all the answers and suggestions :bow: